ipa-core/src/protocol/ipa_prf/shuffle/sharded.rs

```diff
@@ -0,0 +1,522 @@
+        .await?;
+
+    // set our shares
+    let ctx = ctx.narrow(&ShuffleStep::PseudoRandomTable);
+    Ok((0..sz)
+        .map(|i| {
+            // This may be confusing as paper specifies Ã and B̃ as independent tables, but
+            // there is really no reason to generate them using unique PRSS keys.
+            let (a, b) = ctx.prss().generate(RecordId::from(i));
+
+            S::new(a, b)
+        })
+        .collect())
+}
+
+/// Sharded shuffle as performed by shards on H2.
+async fn h2_shuffle<I, S, C>(ctx: C, shares: I) -> Result<Vec<S>, crate::error::Error>
+where
+    I: IntoIterator<Item = S>,
+    I::IntoIter: Send + ExactSizeIterator,
+    C: ShardedContext,
+    S: Shuffleable,
+{
+    // Generate Y_1 = perm_12(right ⊕ z_12)
+    let y1 = ctx
+        .narrow(&ShuffleStep::Permute12)
+        .mask_and_shuffle(
+            Direction::Left,
+            shares.into_iter().map(|share| share.right()),
+        )
+        .await?;
+
+    // Share y1 to the right. Safe to do because input has been masked with randomness
+    // known to H1 and H2 only.
+    ctx.narrow(&ShuffleStep::LeftToRight)
+        .send_all(y1, Direction::Right)
+        .await?;
+
+    let x2 = ctx
+        .narrow(&ShuffleStep::LeftToRight)
+        .recv_all::<S::Share>(Direction::Left)
+        .await?;
+
+    // generate X_3 = perm_23(x_2 ⊕ z_23)
+    let x3 = ctx
+        .narrow(&ShuffleStep::Permute23)
+        .mask_and_shuffle(Direction::Right, x2)
+        .await?;
+
+    // at this moment we know the cardinality of C, and we let H1 know it, so it can start
+    // setting up its own shares.
+    ctx.narrow(&ShuffleStep::Cardinality)
+        .send_word(Direction::Left, x3.len())
+        .await?;
+
+    if x3.is_empty() {
+        return Ok(Vec::new());
+    }
+
+    // Generate c_1 = x_3 ⊕ b, stream it to H3 and receive c_2 from it at the same time.
+    // Knowing b, c_1 and c_2 lets us set our resulting share, according to the paper it is
+    // (b, c_1 + c_2)
+    let send_channel = ctx
+        .narrow(&ShuffleStep::C)
+        .set_total_records(x3.len())
+        .send_channel(ctx.role().peer(Direction::Right));
+    let recv_channel = ctx
+        .narrow(&ShuffleStep::C)
+        .recv_channel(ctx.role().peer(Direction::Right));
+
+    Ok(ctx
+        .try_join(x3.into_iter().enumerate().map(|(i, x3)| {
+            let record_id = RecordId::from(i);
+            let (b, _): (S::Share, S::Share) = ctx
+                .narrow(&ShuffleStep::PseudoRandomTable)
+                .prss()
+                .generate(RecordId::from(i));
+            let c1 = x3 + b.clone();
+            try_join(
+                send_channel.send(record_id, c1.clone()),
+                recv_channel.receive(record_id),
+            )
+            .map_ok(|((), c2)| S::new(b, c1 + c2))
+        }))
+        .await?)
+}
+
+/// Sharded shuffle as performed by shards on H3. Note that H3 does not use its input, it is
+/// dropped immediately.
```

Comment:
Malicious security will change that slightly since we need to generate a MAC using a random MAC key and multiply it with all records (h1, h2 need h3 for that), afterwards h3 can drop them.

Reply:
good catch, let me fix the comment